File: typescript/src/audit-backend.ts
/**
 * Pluggable audit backend abstraction.
 *
 * Defines the AuditBackend interface and provides two implementations:
 *
 * 1. LegacyBackend -- wraps the existing AuditLogger (default).
 * 2. TrailProofBackend -- delegates to TrailProof for tamper-evident event logging.
 *
 * The backend is selected via `audit.backend` in attesta.yaml:
 *
 *   audit:
 *     backend: trailproof  # or "legacy" (default)
 *     path: events.jsonl
 *     tenantId: my-org
 */

import type { ActionContext, ApprovalResult } from "./types.js";
import { Verdict } from "./types.js";
import {
  AuditLogger,
  buildEntry,
  type AuditEntryData,
  type ChainVerification,
} from "./audit.js";

// ---------------------------------------------------------------------------
// Interface
// ---------------------------------------------------------------------------

/**
 * Pluggable audit storage backend.
 *
 * Any object implementing this interface can serve as an audit backend.
 */
export interface AuditBackend {
  /** Persist an audit record and return a unique entry ID. */
  log(ctx: ActionContext, result: ApprovalResult): Promise<string>;

  /** Verify chain integrity. Returns verification result. */
  verify(): ChainVerification;

  /** Return entries matching the supplied filters. */
  query(filters?: Record<string, unknown>): unknown[];
}

// ---------------------------------------------------------------------------
// Legacy backend (wraps existing AuditLogger)
// ---------------------------------------------------------------------------

/**
 * Wraps the existing AuditLogger.
 *
 * This is the default backend when no `audit.backend` is configured.
 * Preserves full backward compatibility with the existing JSONL audit log.
 */
export class LegacyBackend implements AuditBackend {
  private readonly _logger: AuditLogger;

  constructor(path: string = ".attesta/audit.jsonl") {
    this._logger = new AuditLogger(path);
  }

  async log(ctx: ActionContext, result: ApprovalResult): Promise<string> {
    return this._logger.log(ctx, result);
  }

  verify(): ChainVerification {
    return this._logger.verifyChain();
  }

  query(filters?: Record<string, unknown>): AuditEntryData[] {
    return this._logger.query(filters ?? {});
  }

  findRubberStamps(options?: {
    maxReviewSeconds?: number;
    minRisk?: string;
  }): AuditEntryData[] {
    return this._logger.findRubberStamps(options);
  }
}

// ---------------------------------------------------------------------------
// TrailProof backend
// ---------------------------------------------------------------------------

/**
 * Delegates audit logging to TrailProof.
 *
 * Requires: `npm install @kyberonai/trailproof`
 *
 * Maps Attesta audit fields to TrailProof's 10-field event envelope:
 * - `agentId` → `actorId`
 * - `sessionId` → `sessionId`
 * - `tenantId` from config → `tenantId`
 * - Event type: `attesta.approval.{verdict}`
 * - All other Attesta-specific fields → `payload`
 */
export class TrailProofBackend implements AuditBackend {
  private readonly _tp: unknown;
  private readonly _tenantId: string;

  constructor(options: {
    path?: string;
    tenantId?: string;
    hmacKey?: string;
  } = {}) {
    const path = options.path ?? ".attesta/audit.jsonl";
    const tenantId = options.tenantId ?? "default";
    const hmacKey = options.hmacKey;

    this._tenantId = tenantId;

    // Dynamic import to keep trailproof optional
    let Trailproof: unknown;
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      Trailproof = require("@kyberonai/trailproof").Trailproof;
    } catch {
      throw new Error(
        "TrailProof is required for the trailproof audit backend. " +
          "Install with: npm install @kyberonai/trailproof"
      );
    }

    const tpOptions: Record<string, unknown> = {
      store: "jsonl",
      path,
    };
    if (hmacKey != null) {
      tpOptions.hmacKey = hmacKey;
    }

    this._tp = new (Trailproof as new (opts: Record<string, unknown>) => unknown)(
      tpOptions
    ) as Record<string, unknown>;
  }

  async log(ctx: ActionContext, result: ApprovalResult): Promise<string> {
    const entry = buildEntry(ctx, result);

    const verdictStr =
      typeof result.verdict === "string" ? result.verdict : String(result.verdict);
    const eventType = `attesta.approval.${verdictStr}`;

    // All Attesta-specific fields go into the opaque payload
    const payload: Record<string, unknown> = {
      actionName: entry.actionName,
      actionDescription: entry.actionDescription,
      riskScore: entry.riskScore,
      riskLevel: entry.riskLevel,
      challengeType: entry.challengeType,
      challengePassed: entry.challengePassed,
      approverIds: entry.approverIds,
      verdict: entry.verdict,
      reviewDurationSeconds: entry.reviewDurationSeconds,
      minReviewMet: entry.minReviewMet,
      interceptedAt: entry.interceptedAt,
      decidedAt: entry.decidedAt,
      executedAt: entry.executedAt,
      environment: entry.environment,
      metadata: entry.metadata,
    };

    const tp = this._tp as Record<string, (...args: unknown[]) => unknown>;
    const event = tp.emit({
      eventType,
      actorId: ctx.agentId ?? "unknown",
      tenantId: this._tenantId,
      payload,
      sessionId: ctx.sessionId ?? undefined,
    }) as { eventId: string };

    return event.eventId;
  }

  verify(): ChainVerification {
    const tp = this._tp as Record<string, () => unknown>;
    const result = tp.verify() as {
      intact: boolean;
      total: number;
      broken: number[];
    };
    return {
      intact: result.intact,
      totalEntries: result.total,
      brokenLinkIndices: [...result.broken],
    };
  }

  query(filters?: Record<string, unknown>): unknown[] {
    const tp = this._tp as Record<string, (f: unknown) => unknown>;
    const tpFilters: Record<string, unknown> = {};

    if (filters?.agentId != null) {
      tpFilters.actorId = filters.agentId;
    }
    if (filters?.limit != null) {
      tpFilters.limit = filters.limit;
    }

    const result = tp.query(tpFilters) as { events: unknown[] };
    return result.events;
  }

  flush(): void {
    const tp = this._tp as Record<string, () => void>;
    tp.flush();
  }
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/**
 * Create an audit backend by name.
 *
 * @param backend - "legacy" (default) or "trailproof"
 * @param options - Backend configuration options
 */
export function createBackend(
  backend: string = "legacy",
  options: {
    path?: string;
    tenantId?: string;
    hmacKey?: string;
  } = {}
): AuditBackend {
  const normalized = backend.trim().toLowerCase();

  if (normalized === "trailproof") {
    return new TrailProofBackend(options);
  } else if (
    normalized === "legacy" ||
    normalized === "default" ||
    normalized === ""
  ) {
    return new LegacyBackend(options.path);
  } else {
    throw new Error(
      `Unknown audit backend: "${backend}". Use "legacy" or "trailproof".`
    );
  }
}

File: python/src/attesta/core/audit_backend.py
"""Pluggable audit backend abstraction.

Defines the :class:`AuditBackend` protocol and provides two implementations:

1. :class:`LegacyBackend` -- wraps the existing :class:`AuditLogger` (default).
2. :class:`TrailProofBackend` -- delegates to TrailProof for tamper-evident event logging.

The backend is selected via ``audit.backend`` in ``attesta.yaml``::

    audit:
      backend: trailproof   # or "legacy" (default)
      path: events.jsonl
      tenant_id: my-org
"""

from __future__ import annotations

import logging
from typing import Any, Protocol, runtime_checkable

from attesta.core.types import ActionContext, ApprovalResult, Verdict

__all__ = ["AuditBackend", "LegacyBackend", "TrailProofBackend", "create_backend"]

logger = logging.getLogger("attesta.audit")


# ---------------------------------------------------------------------------
# Protocol
# ---------------------------------------------------------------------------


@runtime_checkable
class AuditBackend(Protocol):
    """Pluggable audit storage backend.

    Any object with ``log``, ``verify``, and ``query`` methods that match
    these signatures can serve as an audit backend.
    """

    async def log(self, ctx: ActionContext, result: ApprovalResult) -> str:
        """Persist an audit record and return a unique entry ID."""
        ...

    def verify(self) -> tuple[bool, int, list[int]]:
        """Verify chain integrity. Returns ``(intact, total, broken_indices)``."""
        ...

    def query(self, **filters: Any) -> list[Any]:
        """Return entries matching the supplied filters."""
        ...


# ---------------------------------------------------------------------------
# Legacy backend (wraps existing AuditLogger)
# ---------------------------------------------------------------------------


class LegacyBackend:
    """Wraps the existing :class:`~attesta.core.audit.AuditLogger`.

    This is the default backend when no ``audit.backend`` is configured.
    Preserves full backward compatibility with the existing JSONL audit log.
    """

    def __init__(self, path: str = ".attesta/audit.jsonl") -> None:
        from attesta.core.audit import AuditLogger

        self._logger = AuditLogger(path=path)

    async def log(self, ctx: ActionContext, result: ApprovalResult) -> str:
        return await self._logger.log(ctx, result)

    def verify(self) -> tuple[bool, int, list[int]]:
        return self._logger.verify_chain()

    def query(self, **filters: Any) -> list[Any]:
        return self._logger.query(**filters)

    def find_rubber_stamps(self, **kwargs: Any) -> list[Any]:
        return self._logger.find_rubber_stamps(**kwargs)


# ---------------------------------------------------------------------------
# TrailProof backend
# ---------------------------------------------------------------------------


class TrailProofBackend:
    """Delegates audit logging to TrailProof.

    Requires: ``pip install trailproof``

    Maps Attesta audit fields to TrailProof's 10-field event envelope:
    - ``agent_id`` → ``actor_id``
    - ``session_id`` → ``session_id``
    - ``tenant_id`` from config → ``tenant_id``
    - Event type: ``attesta.approval.{verdict}``
    - All other Attesta-specific fields → ``payload``
    """

    def __init__(
        self,
        *,
        path: str = ".attesta/audit.jsonl",
        tenant_id: str = "default",
        hmac_key: str | None = None,
    ) -> None:
        try:
            from trailproof import Trailproof
        except ImportError:
            raise ImportError(
                "TrailProof is required for the trailproof audit backend. Install with: pip install attesta[trailproof]"
            ) from None

        kwargs: dict[str, Any] = {"store": "jsonl", "path": path}
        if hmac_key is not None:
            kwargs["hmac_key"] = hmac_key

        self._tp = Trailproof(**kwargs)
        self._tenant_id = tenant_id

    async def log(self, ctx: ActionContext, result: ApprovalResult) -> str:
        """Emit a TrailProof event from Attesta audit data."""
        from attesta.core.audit import build_entry

        entry = build_entry(ctx, result)

        verdict_str = result.verdict.value if isinstance(result.verdict, Verdict) else str(result.verdict)
        event_type = f"attesta.approval.{verdict_str}"

        # All Attesta-specific fields go into the opaque payload
        payload = {
            "action_name": entry.action_name,
            "action_description": entry.action_description,
            "risk_score": entry.risk_score,
            "risk_level": entry.risk_level,
            "challenge_type": entry.challenge_type,
            "challenge_passed": entry.challenge_passed,
            "approver_ids": entry.approver_ids,
            "verdict": entry.verdict,
            "review_duration_seconds": entry.review_duration_seconds,
            "min_review_met": entry.min_review_met,
            "intercepted_at": entry.intercepted_at,
            "decided_at": entry.decided_at,
            "executed_at": entry.executed_at,
            "environment": entry.environment,
            "metadata": entry.metadata,
        }

        event = self._tp.emit(
            event_type=event_type,
            actor_id=ctx.agent_id or "unknown",
            tenant_id=self._tenant_id,
            payload=payload,
            session_id=ctx.session_id or None,
        )

        return event.event_id

    def verify(self) -> tuple[bool, int, list[int]]:
        """Verify the TrailProof chain integrity."""
        result = self._tp.verify()
        return (result.intact, result.total, list(result.broken))

    def query(self, **filters: Any) -> list[Any]:
        """Query TrailProof events with Attesta-compatible filters."""
        tp_filters: dict[str, Any] = {}

        # Map Attesta filter keys to TrailProof query params
        if "agent_id" in filters:
            tp_filters["actor_id"] = filters["agent_id"]

        if "limit" in filters:
            tp_filters["limit"] = filters["limit"]

        result = self._tp.query(**tp_filters)
        return list(result.events)

    def flush(self) -> None:
        """Ensure all data is flushed to disk."""
        self._tp.flush()


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


def create_backend(
    backend: str = "legacy",
    *,
    path: str = ".attesta/audit.jsonl",
    tenant_id: str = "default",
    hmac_key: str | None = None,
) -> LegacyBackend | TrailProofBackend:
    """Create an audit backend by name.

    Parameters
    ----------
    backend:
        ``"legacy"`` (default) or ``"trailproof"``.
    path:
        File path for the JSONL audit log.
    tenant_id:
        Tenant identifier for TrailProof backend.
    hmac_key:
        Optional HMAC signing key for TrailProof backend.
    """
    backend = backend.strip().lower()

    if backend == "trailproof":
        return TrailProofBackend(path=path, tenant_id=tenant_id, hmac_key=hmac_key)
    elif backend in ("legacy", "default", ""):
        return LegacyBackend(path=path)
    else:
        raise ValueError(f"Unknown audit backend: {backend!r}. Use 'legacy' or 'trailproof'.")
